### PR TITLE
fix : use RoomMembershipObserver to close room screen when leaving

### PR DESCRIPTION
**Pull request by ganfra**
## Content
Close the room flow when leaving a room from the client.
If leaving, being banned from another client, the room screen will update but won't navigate up to the room list.

## Screenshots / GIFs
No ui change.

## Tests

- Step 1
- Step 2
- Step ...

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request title will be used in the release note, it clearly define what will change for the user
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR

### CODE REVIEW

File: appnav/src/main/kotlin/io/element/android/appnav/room/RoomFlowNode.kt [hunk 1]
```diff
@@ -161,6 +161,14 @@ class RoomFlowNode @AssistedInject constructor(
                 }
             }
         }.launchIn(lifecycleScope)
+
+        // If the user leaves the room from this client, close the room flow.
+        membershipObserver.updates
+            .filter { it.roomId == roomId && !it.isUserInRoom }
+            .onEach {
```
Comment (bmarty) [inline appnav/src/main/kotlin/io/element/android/appnav/room/RoomFlowNode.kt:168]: maybe add `.take(1)` to `navigateUp` only once?

File: features/leaveroom/impl/src/main/kotlin/io/element/android/features/leaveroom/impl/LeaveRoomPresenter.kt [hunk 2]
```diff
@@ -88,20 +85,18 @@ private suspend fun showLeaveRoomAlert(
...
     getRoom(roomId)?.use { room ->
-        room.leave().onSuccess {
-            roomMembershipObserver.notifyUserLeftRoom(room.roomId)
-        }.onFailure {
-            Timber.e(it, "Error while leaving room ${room.displayName} - ${room.roomId}")
-            error.value = LeaveRoomState.Error.Shown
-        }
+        room.leave()
+            .onFailure {
+                Timber.e(it, "Error while leaving room ${room.displayName} - ${room.roomId}")
```
Comment (bmarty) [inline features/leaveroom/impl/src/main/kotlin/io/element/android/features/leaveroom/impl/LeaveRoomPresenter.kt:97]: Not new but we should only log the `roomId` for privacy reason.

**Review by bmarty — APPROVED** [review]
LGTM, 2 small remarks.

File: appnav/src/main/kotlin/io/element/android/appnav/room/RoomFlowNode.kt (hunk 1 above)
Comment (ganfra) [reply appnav/src/main/kotlin/io/element/android/appnav/room/RoomFlowNode.kt:168]: If we get two membership updates there is something wrong, but yeah sure, I can change

File: features/leaveroom/impl/src/main/kotlin/io/element/android/features/leaveroom/impl/LeaveRoomPresenter.kt (hunk 2 above)
Comment (ganfra) [reply features/leaveroom/impl/src/main/kotlin/io/element/android/features/leaveroom/impl/LeaveRoomPresenter.kt]: Changed

File: features/joinroom/impl/src/main/kotlin/io/element/android/features/joinroom/impl/JoinRoomNode.kt [hunk 3]
```diff
@@ -43,14 +43,14 @@ class JoinRoomNode @AssistedInject constructor(
             state = state,
             onBackClick = ::navigateUp,
             onJoinSuccess = ::navigateUp,
-            onCancelKnockSuccess = ::navigateUp,
-            onKnockSuccess = { },
+            onCancelKnockSuccess = {},
+            onKnockSuccess = {},
```
Comment (bmarty) [inline features/joinroom/impl/src/main/kotlin/io/element/android/features/joinroom/impl/JoinRoomNode.kt:47]: Maybe for later, but I believe that the parameters `onCancelKnockSuccess` and `onKnockSuccess` could be removed.

**Review by bmarty — APPROVED** [review]
Thanks for the update!

**Pull request merged**